services/py-api/src/service/mail_service/mail_service.py

```diff
@@ -0,0 +1,63 @@
+from abc import ABC, abstractmethod
+from typing import Literal
+
+
+class MailService(ABC):
+    @abstractmethod
+    async def send_email(
+        self, receiver: str, subject: str, body_content: str, content_type: Literal["text", "html"] = "text"
+    ) -> None:
+        """
+        Sends an email to the specified receiver with the provided subject and body content.
+        Supports both text and HTML bodies.
+        :param receiver: Recipient's email address
+        :param subject: Email subject line
+        :param body_content: Body of the email, could be html or just plain text, must be specified in content_type
+        :param content_type: Defines the type of the email body, could be "html" (for html body) or "text" (for a plain text body)
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    async def send_participant_successful_registration_email(
```

Comment:
What does this have to do with the mailing service? 

You should have a concrete method which is handling the send operation and then you can create multiple different mailing services by extending the base concrete class which would handle the different business domains such as HackathonMailService which will be implementing the `send_participant_successful_registration_email`

Reply:
It's perfectly fine to have an interface/abstract class for the MailService but you generally don't need it as we don't have expect to have different mailing clients - only resend.

Reply:
MailService to be a concrete class with a concrete send method. 

HackathonMailService to extend MailService and implement all domain specific methods. No abstractions is needed whatsoever.

Reply:
> It's perfectly fine to have an interface/abstract class for the MailService but you generally don't need it as we don't expect to have different mailing clients - only resend.

We want to have an abstract class so that we have a well defined interface for sending email which is independent of the mailing service we are using. It is true that for this year we are going to use Resend, however thinking about the future we want to be able to swap the mailing client without causing other necessary changes in our code base. 
As I believe in your understanding of this matter, I think you meant that we are over complicating it. Which to be honest for someone who has taken a simple course in object oriented programming should not be a problem to understand

Reply:
> MailService to be a concrete class with a concrete send method.
> 
> HackathonMailService to extend MailService and implement all domain specific methods. No abstraction is needed whatsoever.

@asynchroza I agree with you on this part. I suggest the following. We can have MailService as an interface and then HacathonMailService implements this interface and adds domain specific methods. 

![image](https://github.com/user-attachments/assets/4657ef57-1419-4563-a30f-90ca9b00a2cb)

Reply:
Well it does not make any sense because the only kind of MailService we will have is a HackathonMailService, the project specifies the domain. This is being overthinked too much, MailService of a hackathon project will specify its projects domain specific methods because we werent going to build a general MailService anyway.

Reply:
I agree with you on the fact that this is the Hackathon API so we should only be sending Hackathon emails. However, it would be a more flexible design if we have a Mail Service class (instead of the interface that I have suggested above), that implements the send_mail() method with a specific mailing client (in this case Resend) and than having more specific classes extending it providing more convenient interfaces (that do not require so many parameters, like the send_participant_successful_email(), ... etc.) for specific usages as you have done. 
This would be super easy to implement now. You just separate the send_mail() form the rest of the methods on a class called Mail Service and have HackathonMailService class extend it and provide the more specific methods that use send_mail() under the hood. @cl3vy let me know what you think.